src/State.hs

```diff
@@ -1299,28 +1324,30 @@ maybeRingBell = do
 -- viewed).  This is a monoid so that it can be folded over when there
 -- are multiple inbound posts to be processed.
 data PostProcessMessageAdd = NoAction
-                           | NotifyUser
+                           | NotifyUser [PostToAdd]
                            | UpdateServerViewed
-                           | NotifyUserAndServer
+                           | NotifyUserAndServer [PostToAdd]
 
 andProcessWith
   :: PostProcessMessageAdd -> PostProcessMessageAdd -> PostProcessMessageAdd
-andProcessWith NotifyUserAndServer _         = NotifyUserAndServer
-andProcessWith _ NotifyUserAndServer         = NotifyUserAndServer
-andProcessWith NotifyUser UpdateServerViewed = NotifyUserAndServer
-andProcessWith UpdateServerViewed NotifyUser = NotifyUserAndServer
-andProcessWith x NoAction                    = x
-andProcessWith _ x                           = x
+andProcessWith NoAction x                                        = x
+andProcessWith (NotifyUserAndServer p) UpdateServerViewed        = NotifyUserAndServer p
+andProcessWith (NotifyUserAndServer p1) (NotifyUser p2)          = NotifyUserAndServer (p1 <> p2)
+andProcessWith (NotifyUserAndServer p1) (NotifyUserAndServer p2) = NotifyUserAndServer (p1 <> p2)
+andProcessWith (NotifyUser p1) (NotifyUser p2)                   = NotifyUser (p1 <> p2)
+andProcessWith (NotifyUser p) UpdateServerViewed                 = NotifyUserAndServer p
+andProcessWith UpdateServerViewed UpdateServerViewed             = UpdateServerViewed
+andProcessWith x y                                               = andProcessWith y x
```

Comment:
This last recursive case that just swaps the order is dangerous, because if someone later adds constructors to the `PostProcessMessageAdd` type but forgets to update the various cases above this one, this can cause an infinite loop. Instead I'd rather have the possible cases enumerated so we can remove this case entirely, because this will cause GHC to warn us when we fail to enumerate the possible cases.